[flutter_tools] Support Integration Tests

jiahaog opened the pull request:
Implements https://flutter.dev/go/tool-integration-test-support

### Usage

Examples below are not exhaustive. Flags compatible with `flutter test` should work as well.

```sh
# Run an Integration Test.
flutter test integration_test/some_test.dart

# Run an Integration Test on the specified device.
flutter -d my-device-id test integration_test/some_test.dart

# Run all tests in `integration_test/` as Integration Tests.
flutter test integration_test
```

Minor changes:

- Translate received test file paths to absolute paths before passing them to the `package:test` runner
- Some tweaking of the `FakeDevice` to support mocking a `LaunchResult`

### Upcoming PR

- Support web
- Device lab tests (#79258) 

Related issue: #66264

jiahaog committed:
[flutter_tools] Support Integration Tests

jiahaog commented on packages/flutter_tools/lib/src/test/test_device.dart:
```diff
@@ -17,7 +16,10 @@ abstract class TestDevice {
   /// Returns a channel that can be used to communicate with the test process.
-  Future<StreamChannel<String>> start({@required String compiledEntrypointPath});
+  ///
+  /// It is up to the device to determine if [entrypointPath] is a precompiled
+  /// or raw source file.
```
Unfortunately the abstraction here isn't very ideal.

For flutter_test, we can accept the precompiled dill, or if we move the compiler into `FlutterTesterTestDevice`, we can accept Dart source files.

For a raw `Device`, it accepts a source file, or a precompiled APK/IPA.

I can't quite think of a way to abstract this better given the awkward compilation of sources that happen in the `FlutterTestPlatform` for the `flutter_tester`. let me know if there's any better ideas.

jonahwilliams replied:
I think its OK, provided you document the convention.

jiahaog replied:
Updated the doc comment for this

jiahaog commented:
@jonahwilliams Let me know if you have any early feedback for this - Most of the problems I've been having here have been resolved. (Still working on adding the tests)

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -201,6 +209,49 @@ class TestCommand extends FlutterCommand {
+      _testFiles = _findTests(testDir).toList();
+      if (_testFiles.isEmpty) {
+        throwToolExit(
+            'Test directory "${testDir.path}" does not appear to contain any test files.\n'
```
nit: indentation

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -201,6 +209,49 @@ class TestCommand extends FlutterCommand {
+    }
+    if (_isIntegrationTest && !_testFiles.every((String file) => file.startsWith('$currentDirectory/integration_test/'))) {
+      throwToolExit(
+        'Integration tests and unit tests cannot be run in a single invocation.'
```
same here

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -23,7 +23,7 @@ import '../test/runner.dart';
 import '../test/watcher.dart';
 
-class TestCommand extends FlutterCommand {
+class TestCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts {
```
Consider extending the description of the test command to describe the difference in behavior between test and integration test commands

jiahaog replied:
Updated the doc comment for this

jonahwilliams commented on packages/flutter_tools/lib/src/test/integration_test_device.dart:
```diff
@@ -0,0 +1,140 @@
+    final vm_service.VmService vmService = await connectToVmService(launchResult.observatoryUri);
+
+    globals.printTrace('test $id: Finding the correct isolate with the integration test service extension');
+    final vm_service.IsolateRef isolateRef = await _findExtensionIsolate(vmService);
```
There may be some timing issues here, since the kIntegrationTestMethod will not be registered until after main finishes, but this may be after the app/isolate has started. There is a waitForExtension method in the resident_runner library that handles this cases, though it doesn't identify the correct Isolate:

https://github.com/flutter/flutter/blob/dea1c46a3c046766089a5d2e85bf6e169411b82e/packages/flutter_tools/lib/src/resident_devtools_handler.dart#L186

jiahaog replied:
Would there be issues with the implementation of `_findExtensionIsolate` method defined below? It looks like it's implementing something similar to the `waitForExtension` method (both subscribing to registered extension events as well as querying for current extensions, which should solve the timing issues here

jonahwilliams replied:
Yeah that makes sense - I would prefer trying to reuse the existing helper if possible.

jiahaog replied:
Refactoring to make the existing helper reusable here was done in #77220, and just updated this PR to use the helper here.

jiahaog committed:
Merge branch 'master' into integration-test

jiahaog committed:
Indentation, docs

jiahaog committed:
Fix an issue with locked devices

jiahaog committed:
Merge branch 'master' into integration-test

jiahaog committed:
Add tests, OS agnostic paths, more exception handling

jiahaog committed:
Merge branch 'master' into integration-test

jiahaog committed:
Additional refactoring, changes to UX

jiahaog committed:
More tests

jiahaog committed:
Fix analyzer

jiahaog committed:
Fix wrong comment

jiahaog commented on packages/flutter_tools/lib/src/test/integration_test_device.dart:
```diff
@@ -0,0 +1,140 @@
+  /// [entrypointPath] must be a path to an uncompiled source file.
+  @override
+  Future<StreamChannel<String>> start(String entrypointPath) async {
+    final TargetPlatform targetPlatform = await device.targetPlatform;
```
Yes, I think we could make `flutter_platform.dart` keep track of some sort of mutable state where rebuilding is skipped for subsequent workflows. For the first run we basically do what is here in this PR, and for subsequent runs we would invoke portions of the attach workflow, to attach and then hot restart the application. I can't be sure if it's sufficient without a prototype, but it seems like a `reloadExistingApp` boolean parameter here would be sufficient. It's also fine if we want to make breaking changes to the interface here because it's all internal anyway.

jonahwilliams replied:
Sounds reasonable to me

jiahaog commented:
Sorry for the delay, I was caught up with some other things, and just found some time to clean this PR up and add more tests. It is finally ready for review 🙂 

(I also have a follow up PR #79258 to add device lab tests)

jonahwilliams reviewed (commented):
Overall this LGTM.

How comfortable do you feel with these changes, do you think we should wait until the next release cut, or until the devicelab test is ready?

As long as the existing `flutter test` behavior` is unchanged it should be OK

jiahaog committed:
Disable port publication so device lab doesn't show the prompt on iOS 14

jiahaog committed:
Use absolute instead of canonical paths to fix windows debugging test

Canonical paths become lowercased on Windows. This causes less
information to be returned from the VM after adding a breakpoint
(observe that the UnresolvedSourceLocation does not have a `script`
below:

bad (at previous commit)

```
{"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"BreakpointAdded","isolate":{"type":"@Isolate","id":"isolates/2153249320239607","name":"main","number":"2153249320239607","isSystemIsolate":false},"timestamp":1617078845364,"breakpoint":{"type":"Breakpoint","fixedId":true,"id":"breakpoints/1","breakpointNumber":1,"resolved":false,"location":{"type":"UnresolvedSourceLocation","scriptUri":"file:///C:/Users/Jia%20Hao/AppData/Local/Temp/flutter_test_expression_eval_test.c001d9de/test/test.dart","line":5}}}}}
```

good (at HEAD)

```
{"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"BreakpointAdded","isolate":{"type":"@Isolate","id":"isolates/2029881360067999","name":"main","number":"2029881360067999","isSystemIsolate":false},"timestamp":1617079201477,"breakpoint":{"type":"Breakpoint","fixedId":true,"id":"breakpoints/1","breakpointNumber":1,"resolved":false,"location":{"type":"UnresolvedSourceLocation","script":{"type":"@Script","fixedId":true,"id":"libraries/@22372424/scripts/file%3A%2F%2F%2FC%3A%2FUsers%2FJia%2520Hao%2FAppData%2FLocal%2FTemp%2Fflutter_test_expression_eval_test.17a2bb2c%2Ftest%2Ftest.dart/178816e6d05","uri":"file:///C:/Users/Jia%20Hao/AppData/Local/Temp/flutter_test_expression_eval_test.17a2bb2c/test/test.dart","_kind":"kernel"},"line":5}}}}}
```

This makes the test [1] confused as it keeps waiting forever.

[1]: https://github.com/flutter/flutter/blob/e43738824e88827984bbd7a85fdb11e0ef5398c9/packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart#L113

jiahaog commented:
I think it should be fine to land this without waiting for the next release cut. The only change to the normal `flutter test` flow is the change to use canonical paths which didn't work on windows when debugging, so I've just updated the PR to use absolute paths. Also just manually tested adding a breakpoint and expression evaluation of the widget test flow on macOS, Linux and Windows in VSCode, and it seems to work fine. Is there anything else I should test manually?

Also, I figured out how to test the device lab changes locally in #79258 (found an issue and updated this PR to disable port publication like how `drive` does it), and we should be good to land the device lab change right after this PR lands. Splitting it off into its own PR so that would be easier to revert, as I'm a bit unsure of whether `dev/prod_builders.json` is specified correctly there.

DanTup commented:
FWIW, I have an [open PR](https://github.com/Dart-Code/Dart-Code/pull/3133) in Dart-Code to support this. I don't plan to merge it for the upcoming release (this week) as I wanted to re-test it all once this PR is complete/merged (it also needs some automated tests). I'll likely merge it just after my release though (and I can make a preview build of the extension that includes the changes for testing/using in the meantime).

jonahwilliams commented on packages/flutter_tools/lib/src/commands/test.dart:
```diff
@@ -201,6 +250,44 @@ class TestCommand extends FlutterCommand {
 
+  @override
+  Future<FlutterCommandResult> verifyThenRunCommand(String commandPath) {
+    _testFiles = argResults.rest.map<String>((String testPath) => globals.fs.path.absolute(testPath)).toList();
```
```suggestion
    _testFiles = argResults.rest.map<String>(globals.fs.path.absolute).toList();
```

jonahwilliams reviewed (approved):
only have a nit.

LGTM

jiahaog committed:
Update packages/flutter_tools/lib/src/commands/test.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

jiahaog commented:
@DanTup sure, no hurry. I'll merge this and then try and patch your change as well to test it out
